core/pkg/monitor/openmetrics.go

```diff
@@ -0,0 +1,279 @@
+package monitor
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/wandb/wandb/core/internal/clients"
+	"github.com/wandb/wandb/core/pkg/observability"
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+
+	"github.com/prometheus/common/expfmt"
+)
+
+// Filter represents a filter to apply to metrics collected from an OpenMetrics endpoint.
+//
+// Provides a fine-grained control over which metrics to collect.
+// OpenMetrics filters can be specified via wandb.Settings in one of the two formats:
+//   - {"metric regex pattern, including endpoint name as prefix": {"label": "label value regex pattern"}}
+//   - ("metric regex pattern 1", "metric regex pattern 2", ...)
+//
+// These are converted to a list of Filter objects using the processMappingFilters
+// and processSequenceFilters functions, respectively.
+type Filter struct {
+	MetricNameRegex string
+	LabelFilters    [][2]string
+}
+
+// processSequenceFilters converts a sequence of metric regex patterns to a list of Filter objects.
+func processSequenceFilters(sequence *spb.ListStringValue) []Filter {
+	if sequence == nil {
+		return nil
+	}
+	result := make([]Filter, len(sequence.GetValue()))
+	for i, filter := range sequence.GetValue() {
+		result[i] = Filter{MetricNameRegex: filter}
+	}
+	return result
+}
+
+// processMappingFilters converts a mapping of metric regex patterns to label regex patterns
+// to a list of Filter objects.
+func processMappingFilters(mapping *spb.MapStringKeyMapStringKeyStringValue) []Filter {
+	if mapping == nil {
+		return nil
+	}
+	result := []Filter{}
+	for metricRegex, labelFilters := range mapping.GetValue() {
+		filter := Filter{MetricNameRegex: metricRegex}
+		for labelName, labelRegex := range labelFilters.GetValue() {
+			filter.LabelFilters = append(filter.LabelFilters, [2]string{labelName, labelRegex})
+		}
+		result = append(result, filter)
+	}
+	return result
+}
+
+// Constants for the default OpenMetrics HTTP client configuration.
+//
+// Metrics are collected on a best-effort basis, but we do allow for some retries.
+const (
+	DefaultOpenMetricsRetryMax     = 3
+	DefaultOpenMetricsRetryWaitMin = 1 * time.Second
+	DefaultOpenMetricsRetryWaitMax = 10 * time.Second
+	DefaultOpenMetricsTimeout      = 5 * time.Second
+)
+
+// OpenMetrics is a monitor that collects metrics from an OpenMetrics endpoint.
+//
+// Supports only GAUGE and COUNTER metrics.
+type OpenMetrics struct {
+	name        string
+	url         string
+	filters     []Filter
+	client      *retryablehttp.Client
+	logger      *observability.CoreLogger
+	labelMap    map[string]map[string]int
+	labelHashes map[string]map[string]string
```

Comment:
some documentation of the different fields would be helpful, specifically:
- filters, labelMap, labelHashes (it's not obvious without reading the code what these are used for)